### PR TITLE
Feat: Add custom select component

### PR DESCRIPTION
**Pull request by lolottetheclash**
This PR implements Select component whose issue is [here](https://github.com/okp4/ui/issues/122).

As we discuss, the options can be with or without groups:

Capture d’écran 2022-07-29 à 14 01 13

It can also allow multiple choices:

Capture d’écran 2022-07-29 à 14 01 48

We also decided to make the component fully controlled, as mentioned in the story.

### CODE REVIEW

**Review by JoeyDjr — APPROVED** [review]
Great job ! 👏 🎊

File: src/ui/molecules/select/Select.tsx [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+export type SelectProps = InputPropsForSelect & {
+  /**
+   * Defines the callback called when the select value changes.
+   */
+  readonly onChange: (value: string | Readonly<string[]>) => void
+  /**
+   * The value of the select.
+   */
+  readonly value: string | string[]
```
Comment (ErikssonJoakim) [inline src/ui/molecules/select/Select.tsx:34]: Consider renaming to something like `selected` to make it more clear what the value is

File: src/ui/molecules/select/Select.tsx [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  size = 'small',
+  disabled = false,
+  hasError = false,
+  multiple = false,
+  options,
+  onChange,
+  fullWidth,
+  value,
+  helperText
+}: SelectProps): JSX.Element => {
```
Comment (ErikssonJoakim) [inline src/ui/molecules/select/Select.tsx:71]: It's up to you but you could wrap the SelectProps in a deepReadonly and get rid of the Readonly in SelectProps. Makes the type a bit easier to read 😄

**Review by ErikssonJoakim — CHANGES_REQUESTED** [review]
Great work! (In both sense of the word 😄 ). I've left some comments to consider, hope you like them 🍝

File: src/ui/molecules/select/Select.tsx (hunk 1 above)
Comment (lolottetheclash) [reply src/ui/molecules/select/Select.tsx:34]: In general the value selected by the select is called `value`, I prefer to keep this usage.

File: src/utils.ts [hunk 3]
```diff
@@ -24,3 +25,58 @@ export const toPercent = (value: number, min: number, max: number): number => {
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export const isString = (value: any): value is string => typeof value === 'string'
+
+export const sort = (value1: string, value2: string): number => value1.localeCompare(value2)
```
[the remark this answers is not in the document]
Comment (lolottetheclash) [reply src/utils.ts:29]: Good point! I had forgotten that it was already implemented!

File: src/utils.ts [hunk 4]
```diff
@@ -24,3 +25,58 @@ export const toPercent = (value: number, min: number, max: number): number => {
...
+
+export const compareStrings = (value1: string, value2: string): number =>
```
[the remark this answers is not in the document]
Comment (lolottetheclash) [reply src/utils.ts:31]: According to me the name of the function is clear enough to express its role, a bit like `localeCompare` where `string` does not appear in its name either. 
However, we can keep `refString` and `compareString` to keep consistency with `localeCompare` method documentation. 
This is a good idea to add locales & options in order to offer more flexibility to this utility function 👍

**Review by ErikssonJoakim — APPROVED** [review]
Awesome work 🚀

File: src/utils.ts [hunk 5]
```diff
@@ -24,3 +25,62 @@ export const toPercent = (value: number, min: number, max: number): number => {
...
+    return -1
+  }
+  return option2.value.localeCompare(option1.value)
+}
+
+export const getOptionsAscendingSorted = (options: Readonly<Option[]>): Option[] => {
+  return [...options].sort((option1: Option, option2: Option) =>
+    ascendingSortByGroupAndValues(option1, option2)
+  )
+}
```
Comment (ccamel) [inline src/utils.ts:72]: Apply  [η-conversion](https://wiki.haskell.org/Eta_conversion) (wink to @fredericvilcot) that allows you to drop the abstraction over a function and get rid of this ugly `return`:

```suggestion
export const getOptionsAscendingSorted = (options: Readonly<SelectOption[]>): SelectOption[] =>
  [...options].sort(ascendingSortByGroupAndValues)
```

File: src/utils.ts [hunk 6]
```diff
@@ -24,3 +25,62 @@ export const toPercent = (value: number, min: number, max: number): number => {
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export const isString = (value: any): value is string => typeof value === 'string'
+
+export const compareStrings = (
```
Comment (ccamel) [inline src/utils.ts:29]: Please provide tests for all the functions introduced hre.

**Review by ccamel — CHANGES_REQUESTED** [review]

File: src/utils.ts [hunk 7]
```diff
@@ -24,3 +25,62 @@ export const toPercent = (value: number, min: number, max: number): number => {
...
+): number => referenceStr.localeCompare(compareString, locales, options)
+
+const ascendingSortByGroupAndValues = (
+  option1: Readonly<Option>,
+  option2: Readonly<Option>
+): number => {
+  const definedOption1 = option1.group ?? ''
+  const definedOption2 = option2.group ?? ''
+
+  if (definedOption1 > definedOption2) {
```
Comment (ccamel) [inline src/utils.ts:45]: If you compare the values of the option using `compareStrings`, you should do the same for the groups.

File: src/utils.ts [hunk 8]
```diff
@@ -24,3 +25,62 @@ export const toPercent = (value: number, min: number, max: number): number => {
...
+  locales?: string | string[],
+  // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
+  options?: Intl.CollatorOptions | undefined
+): number => referenceStr.localeCompare(compareString, locales, options)
+
+const ascendingSortByGroupAndValues = (
+  option1: Readonly<Option>,
+  option2: Readonly<Option>
+): number => {
+  const definedOption1 = option1.group ?? ''
```
Comment (ccamel) [inline src/utils.ts:42]: Should be named:
```suggestion
  const group1 = option1.group ?? ''
```

File: src/utils.ts [hunk 9]
```diff
@@ -24,3 +25,62 @@ export const toPercent = (value: number, min: number, max: number): number => {
...
+const descendingSortByGroupAndValues = (option1: Option, option2: Option): number => {
+  const definedOption1 = option1.group ?? ''
+  const definedOption2 = option2.group ?? ''
+  if (definedOption1 < definedOption2) {
+    return 1
+  }
+  if (definedOption1 > definedOption2) {
+    return -1
+  }
+  return option2.value.localeCompare(option1.value)
```
Comment (ccamel) [inline src/utils.ts:65]: Why using a `compareStrings` for ascending, and not the same function for descending?

**Comment by ccamel** [comment]
@lolottetheclash @fredericvilcot I pushed some modifications to the code, some refactoring, and I introduced 2 hooks to address possible transversal concerns. There are still many things to improve such as the use of a memo for the computation of data and we should also answer to the question of whether the component should handle the capitalization of labels. In my opinion no, this should be provided by the caller, in which case we could delete the corresponding code.

**Review by ccamel — APPROVED** [review]

**Review by JoeyDjr — APPROVED** [review]

**Review by fredericvilcot — APPROVED** [review]
@lolottetheclash That's a great piece of work and the component looks sexy, thanks for that! 🚀
We pushed with @ccamel some commits to refactor code, it remains some minor element to refactor that we can discuss later!
Let's merge it!

**Review by ccamel — APPROVED** [review]

**Pull request merged**